### PR TITLE
Make Matrix#[]= public method

### PR DESCRIPTION
**Pull request by grzegorz-jakubiak**
To address this issue [#14151](https://bugs.ruby-lang.org/issues/14151)

### CODE REVIEW

**Comment by greggzst** [comment]
@marcandre Ping

**Comment by marcandre** [comment]
As is, this PR is not acceptable. Here's the todo list for this PR to be make Matrix mutable in an acceptable way:

1) deal with indices as ranges

`m[1..2, 1..2] = 42` #  => Sets the 4 entries to 42. Must accept small negative ranges too
`m[1..2, 1..2] = matrix` # => Sets the 4 entries from the entries of the given 2x2 matrix
Same with  `m[1, 1..2] = 42 / vector / matrix` and `m[1..2, 1]=...`

Note: it's acceptable to use `is_a?(Range)` to recognize ranges. For ints, there's already  `coerce_to_int`.

2) Define `map!`, which should accept a parameter `:diagonal`, etc.

3) Handle `freeze` correctly.
`m.freeze.[...]= ...` must raise, same with `map!`

4) `dup` probably needs to be defined

5) specs for all of these, ideally in `ruby/spec`

Let me know if you're up to update your PR or not...

**Comment by greggzst** [comment]
@marcandre Thanks for your update and feedback. At the moment I'm unable to update my PR to meet all of the things you mentioned.(work and school related) But I'll be able to do it in the near future unless there is some specific schedule or anybody else wants to do it.

**Comment by greggzst** [comment]
@marcandre I took care of the first point you mentioned. I'd be grateful if you or somebody could provide some feedback about my code so I could fix or simplify it. Shortly, I'll create commits with regard to the rest of points.

File: lib/matrix.rb [hunk 1]
```diff
@@ -304,11 +304,87 @@ def [](i, j)
...
+        end
+      else
+        j.each do |j|
+          @rows[i][j] = v
+        end
+      end
+    else
+      CoercionHelper.coerce_to_int(i)
+      CoercionHelper.coerce_to_int(j)
+      raise ArgumentError, "indecies are outside of matrix" unless row_range.cover?(i) && column_range.cover?(j)
```
Comment (Maumagnaguagno) [inline lib/matrix.rb]: Replace "indecies" with "indices" or "indexes".

**Comment by Maumagnaguagno** [comment]
You could replace ``is_subrange?(row_range, i)`` with an optimized ``in_row_range?(i)`` that avoids creating a temporary Range, the same could be done for ``column_range``.
The current method ``is_subrange?`` looks like a good candidate for a Range method.
Maybe renamed to ``subrange?`` or ``between?`` if moved to Range, to match ``cover?`` and ``include?``.

The calls to ``row_range.cover?(i)`` and ``column_range.cover?(j)`` could be replaced by ``i.between?(-row, row-1)`` or just ``-row <= i and i < row``, as they are used only once each.
I don't see the need for ``row_range`` and ``column_range`` as temporary ``-x...x`` range creators.

**Comment by greggzst** [comment]
@Maumagnaguagno Thanks a lot for your feedback. I made changes according to it.

**Comment by Maumagnaguagno** [comment]
Hey @greggzst, I noticed you fixed the last commit with a ``range.to_a`` to fix exclude end Ranges.
This works, but I believe it is slow, as it creates an array with every element in range.
Instead you could just use ``range.max`` to pick the last element in range.

**Comment by greggzst** [comment]
@Maumagnaguagno Thanks! Fixed that!

**Comment by marcandre** [comment]
Lots of good work in there.

I notice right away a few things:
- there are some `range.to_a` left. You shouldn't ever have to call `to_a`.
- Use `i = coerce_to_int(i)` (i.e. use the result). If you were writing ruby-spec style instead of tests, this would have been more easier to catch.
- `IndexError` is the more precise exception class to raise for bad indices
- It's great you have tests; could you split them a bit? It's hard to follow what's tested (and what might be missing)

**Comment by greggzst** [comment]
@marcandre Thanks I lot I addressed every point but `range.to_a`. Is it acceptable to use `each_with_index` on range to get the appropriate index from it?

```
v.each_with_index do |e, row, col|
  r = nil
  i.each_with_index do |i , index|
    if index == row
      r = i
      break
    end
  end
  c = nil
  j.each_with_index do |j , index|
    if index == col
      c = j
      break
    end
  end
  @rows[r][c] = e
end
```
I can't come up with any other idea and couldn't find any method in `Range` that returns element at specific index.

**Comment by Maumagnaguagno** [comment]
Maybe ``r = i.first + row``?

**Comment by greggzst** [comment]
I definied `map!` and `freeze` any feedback about these two?

**Comment by greggzst** [comment]
Implemented all things for `Vector` as well but there is no `IndexError` defined. Should I define it?

File: lib/matrix.rb [hunk 2]
```diff
@@ -445,11 +445,17 @@ def collect(&block) # :yield: e
   alias map collect
 
   def collect!(which = :all)
+    return to_enum(:collect!) unless block_given?
```
Comment (marcandre) [inline lib/matrix.rb]: should be `to_enum(:collect!, which)`

File: test/matrix/test_matrix.rb [hunk 3]
```diff
@@ -684,6 +684,20 @@ def test_set_element
     assert_raise(Matrix::ErrDimensionMismatch){ m1[1, 0..3] = Matrix[[1],[5]] }
   end
 
+  def test_map!
+    m1 = Matrix.zero(2,2)
```
Comment (marcandre) [inline test/matrix/test_matrix.rb]: Need a test that the result is the receiver.

**Comment by greggzst** [comment]
All done anything else I should do?

File: lib/matrix.rb [hunk 4]
```diff
@@ -1750,11 +1834,39 @@ def [](i)
   alias component []
 
   def []=(i, v)
-    @elements[i]= v
+    if i.is_a?(Range)
+      raise IndexError 'expected range is outside of vector' unless in_vector_range?(i)
```
Comment (Maumagnaguagno) [inline lib/matrix.rb]: Missing comma between IndexError and message.

Comment (greggzst) [reply lib/matrix.rb]: Totally missed that. Thanks a lot!

**Comment by greggzst** [comment]
The message associated with `FrozenError` is wrong and it says `can't modify frozen Array` is there any way to override it?

**Comment by greggzst** [comment]
I think that the way `clone` method was implemented both in `Matrix` and `Vector` is actually the way `dup` should be implemented because both of them disregard state of the object. So I renamed them to `dup` and defined proper `map` for `Matrix` which accepts parameters through `which` as other methods.

**Comment by greggzst** [comment]
@marcandre Any feedback on `map` implementation? What about the message produced by `FrozenError`? It's currently wrong as I mentioned it earlier.

**Comment by greggzst** [comment]
@marcandre ping

**Comment by marcandre** [comment]
Sorry, I'm travelling right now.

`map!` looks pretty good at first sight. It's probably worth it to use `@rows[row_index][col_index] = ...` instead of `self[row_index, col_index] = ...`, though. Less method calls, range checking, etc.

Also, should raise an error on empty matrices (like `[].map!{}` does). Good opportunity to fix the message?

**Comment by greggzst** [comment]
I'm using `@rows` right now. `[].map!{}` doesn't actually raise any error(checked it in IRB and it returns `[]`) so I didn't change it for matrices. When it comes to `FrozenError` I still don't know how to fix the message but I think the main issue is that it is invoked when `@rows[row][col] = ...` is called and that's the reason why the message says `can't modify frozen Array` instead of `can't modify frozen Matrix`. Any suggestions on that? Also I fixed conflict on this branch by creating a new one from recent `trunk` and `cherry-picking` my commits.

**Comment by marcandre** [comment]
Sorry, I meant `[].freeze.map!{}`, which does raise.

**Comment by greggzst** [comment]
Okay now I see but I can't get my head around this I mean I don't know how to fix it for know because with empty matrix it never gets to the inner `each_with_index` where `yield` gets called and that's why the `FrozenError` isn't raised. I'll try to figure it out later on but if you've got any tips on that let me know.

I think we don't have to care about `FrozenError` message anymore I checked `Set` and its `FrozenError` messages say `can't modify frozen Hash` so I guess its fine.

**Comment by marcandre** [comment]
Indeed. You'll need to add an explicit `raise .... if frozen?`

**Comment by greggzst** [comment]
I introduced some refactoring and optimizations I came up with. I benchmarked most of them as well using this code:
```
require 'benchmark/ips'
require 'matrix'

a = Array.new(1000){ Array.new(1000){ rand(25..1250) } }
i = 50..980
j = 20..700

Benchmark.ips do |x|
  x.report('set_row_and_col - double each') do
    i.each do |i|
      j.each do |j|
        a[i][j] = 5
      end
    end
  end

  x.report('set_row_and_col - one each and range - Array.new') do
    set_value = Array.new(j.size, 5)
    i.each do |i|
      a[i][j] = set_value
    end
  end

  x.report('set_row_and_col - one each and range - map!') do
    i.each do |i|
      a[i][j].map!{ 5 }
    end
  end

  x.compare!
end

i = 200
j = 20..800

Benchmark.ips do |x|
  x.report('set_col_range - each') do
    j.each do |j|
      a[i][j] = 34
    end
  end

  x.report('set_col_range - no each - Array.new') do
    a[i][j] = Array.new(j.size, 34)
  end


  x.report('set_col_range - map!') do
    a[i][j].map!{ 34 }
  end

  x.compare!
end

i = 1..800
j = 6

Benchmark.ips do |x|
  x.report('set_row_range - each row') do
    a[i].each do |i|
      i[j] = 34
    end
  end

  x.report('set_row_range - no each - map') do
    a[i].map!{|e| e[j] = 34}
  end

  x.compare!
end

m = Matrix[[1,2,3,5,6]]
v = [4,5,6,7,8,9,10,11,12,13,34]
i = 0...5

Benchmark.ips do |x|
  x.report('set_range - matrix - each') do
    m.each_with_index do |e, row, col|
      r = i.first + col
      v[r] = e
    end
  end

  x.report('set_range - matrix - to_a') do
    v[i] = m.to_a[0]
  end

  x.compare!
end

m = Matrix.build(0,1000){ |row, col| (row - col) * -13 }
rows = Array.new(1000) { Array.new(2000) { rand(12..1567) }}
i = 0..700
j = 1000

Benchmark.ips do |x|
  x.report('set_row_range to column matrix - each_with_index') do
    m.each_with_index do |e, row, col|
      r = i.first + row
      rows[r][j] = e
    end
  end

  x.report('set_row_range to column matrix - map!') do
    index = i.first
    mat = m.column(0).to_a
    rows[i].map!{|e| e[j] = mat[index]; index = index + 1 }
  end

  x.compare!
end
```

These are the outcomes:
```
Comparison:
set_row_and_col - one each and range - Array.new:     1475.0 i/s
set_row_and_col - one each and range - map!:       26.0 i/s - 56.83x  slower
set_row_and_col - double each:       19.9 i/s - 73.94x  slower

Comparison:
set_col_range - no each - Array.new:   462402.6 i/s
set_col_range - map!:    21274.2 i/s - 21.74x  slower
set_col_range - each:    14141.7 i/s - 32.70x  slower

Comparison:
set_row_range - each row:    14547.4 i/s
set_row_range - no each - map:    13354.5 i/s - same-ish: difference falls within error

Comparison:
set_range - matrix - to_a:  1260172.4 i/s
set_range - matrix - each:   501323.9 i/s - 2.51x  slower

Comparison:
set_row_range to column matrix - each_with_index:  1713121.2 i/s
set_row_range to column matrix - map!:    12714.0 i/s - 134.74x  slower
```

**Comment by greggzst** [comment]
@marcandre @Maumagnaguagno  any feedback on changes I've made?

**Comment by marcandre** [comment]
Good work.

One thing I notice: ranges like 2..-1 should be supported.

Style-wise, I'm not super enthusiastic at having so many private methods. For normal gems, I tend to do this, but standard lib style is usually "simpler" (for example with all the code in the same file) and I'd rather have the minimum number of private methods possible. If you want to keep the `raise_...` methods, they have to be renamed to something that describes better what they are doing (e.g. `check_...` or `insure_...`)

**Comment by greggzst** [comment]
I can't seem to understand how these ranges are supposed to work. I mean when I call `size` on `2..-1` I get `0` so I prepared my own way of getting the size of these ranges but I see it's wrong because `2..-1` doesn't correspond to array like this `[2, 1, 0, -1]`.  My example:
```
a = [[1, 2, 3], [4, 5, 6], [7, 8, 9]] 
=> [[1, 2, 3], [4, 5, 6], [7, 8, 9]] 
i = 2..-1
=> 2..-1
a[1][2..-1]
=> [6]
```
@marcandre could you explain how these ranges work and how I should tackle them?

**Comment by marcandre** [comment]
You'll probably have to call `.begin`, `.end` and `.exclude_end?`

**Comment by greggzst** [comment]
I know that but the way I'm calculating size is wrong and I don't really understand how this should work. That's my current code:
```
def range_size(range)
    first = range.first
    last = range.last
    if first > last
      if range.exclude_end?
        first - (last + 1) + 1
      else
        first - last + 1
      end
    else
      range.size
    end
  end
```
and when I execute this code:
```
m11 = Matrix[[1, 2, 3], [4, 5, 6], [7, 8, 9]]
m11[1, 2..-1] = 78
Matrix[[1, 2, 3], [78, 78, 6], [7, 8, 9]] == m11
```
I get false because m11 becomes `Matrix[[1, 2, 3], [4, 5, 78, 78, 78, 78], [7, 8, 9]]` which is wrong because as I said I don't really know how these ranges work and why `2..-1` isn't equal to `2,1,0,-1` but only to `2,-1` or something? Could you at least shed some light on that?

**Comment by marcandre** [comment]
`(a .. -2) <=> (a .. (size - 2) )`, so use `size + end - begin (+1?)`

**Comment by greggzst** [comment]
@marcandre done! I'm using `convert_range` to handle these ranges which doesn't require fixing every each and so on. I changed one thing in `set_row_and_col_range` because I noticed a mistake with setting matrices so I used the previous code fragment that works 100%.

**Comment by Maumagnaguagno** [comment]
Both methods `` range_within_count?`` and ``range_within_vector_range?`` use ``max <= (var - 1)`` which is equivalent to ``max < var``.

**Comment by greggzst** [comment]
Thanks! Already corrected.

**Comment by greggzst** [comment]
@marcandre any thoughts or feedback? I reckon everything is fine and ready for merge.

**Comment by greggzst** [comment]
@marcandre ping

File: lib/matrix.rb [hunk 5]
```diff
@@ -303,12 +303,139 @@ def [](i, j)
...
+      value_to_set = Array.new(col_range.size, value)
+      row_range.each do |i|
+        @rows[i][col_range] = value_to_set
+      end
+    end
+  end
+
+  def set_row_range(row_range, col, value)
+    if value.is_a?(Vector)
+      raise ArgumentError, "vector to be set has wrong size" unless row_range.size == value.size
```
Comment (marcandre) [inline lib/matrix.rb]: Best raise a DimensionError here.

File: lib/matrix.rb [hunk 6]
```diff
@@ -303,12 +303,139 @@ def [](i, j)
...
+      set_column_vector(row_range, col, value)
+    elsif value.is_a?(Matrix)
+      Matrix.Raise ErrDimensionMismatch unless value.column_count == 1
+      set_column_vector(row_range, col, value.column(0))
+    else
+      @rows[row_range].map!{|e| e[col] = value }
```
Comment (marcandre) [inline lib/matrix.rb]: `map!` is wrong here. There has to be a missing test too.

File: lib/matrix.rb [hunk 7]
```diff
@@ -303,12 +303,139 @@ def [](i, j)
...
+    end
+  end
+
+  def set_column_vector(row_range, col, value)
+    value.each_with_index do |e, index|
+      r = row_range.first + index
```
Comment (marcandre) [inline lib/matrix.rb]: Best be consistent and use `begin`

File: lib/matrix.rb [hunk 8]
```diff
@@ -303,12 +303,139 @@ def [](i, j)
...
+      raise IndexError, "expected column range is outside of matrix" unless range_within_count?(col, column_count)
+      set_col_range(row, col, value)
+    end
+  end
+
+  def set_row_and_col_range(row_range, col_range, value)
+    if value.is_a?(Matrix)
+      Matrix.Raise ErrDimensionMismatch unless ranges_and_dimensions_equal?(row_range, col_range, value)
+      value.each_with_index do |e, row, col|
+        r = row_range.first + row
```
Comment (marcandre) [inline lib/matrix.rb]: Use `begin`, next line too.

File: lib/matrix.rb [hunk 9]
```diff
@@ -361,17 +488,49 @@ def column(j) # :yield: e
...
+  def collect!(which = :all)
+    return to_enum(:collect!, which) unless block_given?
+    raise FrozenError, "can't modify frozen Matrix" if frozen?
+    each_with_index(which){ |e, row_index, col_index| @rows[row_index][col_index] = yield e }
+  end
+
+  alias map! collect!
+
+  def freeze
+    @rows.map!(&:freeze).freeze
```
Comment (marcandre) [inline lib/matrix.rb]: This won't work if calling `freeze` twice on the same object. Test missing too.

File: test/matrix/test_matrix.rb [hunk 10]
```diff
@@ -623,6 +634,120 @@ def test_combine
     assert_equal Matrix.empty(0,3), Matrix.combine(Matrix.empty(0,3), Matrix.empty(0,3)) { raise }
   end
 
+  def test_set_element
+    m1 = Matrix[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]
```
Comment (marcandre) [inline test/matrix/test_matrix.rb]: Tests should be in the same order as the doc.
Successive tests should be independent.
Tests would be more readable if the matrices used were closer to the `[]=` instead of at the beginning of method.

File: lib/matrix.rb [hunk 11]
```diff
@@ -303,12 +303,139 @@ def [](i, j)
...
+
+  private :indices_within_matrix?, :both_ranges?, :one_is_range?, :convert_range
+
+  def set_value(row, col, value)
+    @rows[row][col] = value
+  end
+
+  def set_range(row, col, value)
+    if row.is_a?(Range)
+      row = convert_range(row, row_count) if row.begin > row.end
```
Comment (marcandre) [inline lib/matrix.rb]: I have the feeling that it would be much better to do the checking and error raising in the same method. Error message could be made more instructive by giving the actual wrong values too.

File: lib/matrix.rb [hunk 12]
```diff
@@ -303,12 +303,139 @@ def [](i, j)
...
+  end
+
+  private :indices_within_matrix?, :both_ranges?, :one_is_range?, :convert_range
+
+  def set_value(row, col, value)
+    @rows[row][col] = value
+  end
+
+  def set_range(row, col, value)
+    if row.is_a?(Range)
```
Comment (marcandre) [inline lib/matrix.rb]: It's a bit ugly that this is the third time we're calling `row.is_a?(Range)`...

File: lib/matrix.rb [hunk 13]
```diff
@@ -303,12 +303,139 @@ def [](i, j)
...
+    else
+      i = CoercionHelper.coerce_to_int(i)
+      j = CoercionHelper.coerce_to_int(j)
+      raise IndexError, "indices are outside of matrix" unless indices_within_matrix?(i, j)
+      set_value(i, j, v)
+    end
   end
   alias set_element []=
   alias set_component []=
-  private :[]=, :set_element, :set_component
```
Comment (marcandre) [inline lib/matrix.rb]: `set_element` and `set_component` should remain private. They are there only for historical reasons.

File: lib/matrix.rb [hunk 14]
```diff
@@ -361,17 +488,49 @@ def column(j) # :yield: e
...
+  # Elements can be restricted by passing an argument:
+  # * :all (default): yields all elements
+  # * :diagonal: yields only elements on the diagonal
+  # * :off_diagonal: yields all elements except on the diagonal
+  # * :lower: yields only elements on or below the diagonal
+  # * :strict_lower: yields only elements below the diagonal
+  # * :strict_upper: yields only elements above the diagonal
+  # * :upper: yields only elements on or above the diagonal
+  #
+  def collect!(which = :all)
```
Comment (marcandre) [inline lib/matrix.rb]: It's good that `map/collect` accept a parameter, but it has to be tested.

Comment (greggzst) [reply lib/matrix.rb]: They are tested: `test_matrxi.rb` lines: `285..298 & 759..777`

**Comment by greggzst** [comment]
@marcandre I think is all done now

File: lib/matrix.rb (hunk 14 above)
Comment (marcandre) [reply lib/matrix.rb]: Oh, indeed, sorry

**Comment by greggzst** [comment]
@marcandre any feedback? After all these fixes it's probably ready to be merged.

**Comment by greggzst** [comment]
@marcandre ping

**Comment by greggzst** [comment]
@marcandre ping

**Comment by marcandre** [comment]
Sorry I haven't had time to look at it. Still in my todo list!

**Comment by greggzst** [comment]
Okay, no worries. I just wanted to remind you.

**Comment by grzegorz-jakubiak** [comment]
@marcandre Have you had time to take a look?

**Comment by grzegorz-jakubiak** [comment]
@marcandre ping

**Comment by grzegorz-jakubiak** [comment]
@marcandre I really want to make sure my code is okay and ready to be merged so as to serve other matrix lovers. So if you've had a chance to look at it just let me know.

**Comment by grzegorz-jakubiak** [comment]
@marcandre ping

**Comment by grzegorz-jakubiak** [comment]
@marcandre any chance you took a look at the code?

**Comment by marcandre** [comment]
I've started looking at it and making final tweaks. Sorry it's taking me so long. In any case the goal is for next major...

**Comment by Maumagnaguagno** [comment]
I just noticed that ``Range.new`` could simplify both ``convert_range`` implementations.
```Ruby
def convert_range(range, count)
  #range.exclude_end? ? range.begin...(count + range.end) : range.begin..(count + range.end)
  Range.new(range.begin, range.end + count, range.exclude_end?)
end
```

**Comment by grzegorz-jakubiak** [comment]
@marcandre Thanks! Will it be merged before December 2018 then? I'd really want to see it as a part of one of the preview versions. BTW I've just pushed new version of `convert_range` suggested by @Maumagnaguagno.

**Comment by marcandre** [comment]
@Maumagnaguagno indeed, already in my local branch :-)

@grzegorz-jakubiak thanks. At this point probably best to wait for my code though.

**Comment by grzegorz-jakubiak** [comment]
@marcandre I didn't get you're gonna improve it. So, I'm gonna wait till you finish but I really hope it's gonna be merged before 2.6 is released.

**Comment by marcandre** [comment]
@grzegorz-jakubiak Haven't forgotten, but focussing on DeepCover and my upcoming talk at RubyKaigi. Hopefully I'll have the time to hack on this on the plane or something, otherwise right after. Yes, it's definitely my objective that this gets merged before next major

**Comment by grzegorz-jakubiak** [comment]
@marcandre Thanks I hope it gets merged before 2.6 release.

**Comment by marcandre** [comment]
No worries. 2.6 is planned for christmas, as usual. I'm on my way back to Montreal too :-)

**Comment by grzegorz-jakubiak** [comment]
@marcandre Thanks, can't wait to use it in new release. If you need any help or anything just let me know

**Comment by marcandre** [comment]
@grzegorz-jakubiak you'll probably be interested in checking #1905!
